### PR TITLE
Stacktrace

### PR DESCRIPTION
**Pull request by Rythmag**
1. StackTrace class finds the call stack of an error given the log line number.
2. StackTraceFormat defines the regex used for finding stacktrace.
3. StackTraceServlet returns the json response for the stacktrace found

### CODE REVIEW

File: src/main/java/com/google/error_analyzer/backend/StackTrace.java [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+                if ( StackTraceFormat.matchesFormat(logText) ) {
+                    logger.info("Adding to stack list log line number "
+                    .concat(logLineNumber.toString()));
+                    stackLogLines.add(logText);
+                } else {
+                    return stackLogLines.build();
+                }
+            }
+        }
+    }
```
Comment (nishantnahata) [inline src/main/java/com/google/error_analyzer/backend/StackTrace.java:133]: Most of it is repetitive (similar to findStack logic). Make it modular.

File: src/main/java/com/google/error_analyzer/servlets/StackTraceServlet.java [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+
+/** Servlet to request for stack trace following an error 
+*/
+@WebServlet("/stackTrace")
+public class StackTraceServlet extends HttpServlet {
+    private static final Logger logger = LogManager.getLogger(StackTraceServlet.class);
+    public StackTrace stackTrace = new StackTrace();
+    @Override
+    public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+        response.setContentType("application/json");
```
Comment (nishantnahata) [inline src/main/java/com/google/error_analyzer/servlets/StackTraceServlet.java:40]: I guess this responseType is in some constants class created by astha

File: src/main/java/com/google/error_analyzer/servlets/StackTraceServlet.java [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+            String json = gson.toJson(stackList);
+            response.getWriter().println(json);
+        } catch (NumberFormatException numberFormatException) {
+            String errorMsg = "Could not parse logLineNumber ".concat(numberFormatException.toString());
+            logger.error(errorMsg);
+            response.getWriter().println(errorMsg);
+        } catch (IOException ioException) {
+            String errorMsg = "Could not connect to database";
+            logger.error(errorMsg);
+            response.getWriter().println(errorMsg);
```
Comment (nishantnahata) [inline src/main/java/com/google/error_analyzer/servlets/StackTraceServlet.java:55]: all the responses should be in json

File: src/test/java/com/google/error_analyzer/backend/StackTraceTest.java [hunk 4]
```diff
@@ -0,0 +1,179 @@
...
+    public void setUp() {
+        stackTrace = new StackTrace();
+        stackTrace.logDao = Mockito.mock(MockLogDao.class);
+        ReflectionTestUtils.setField(stackTrace, "BATCH_SIZE", BATCH_SIZE);
+        ReflectionTestUtils.setField(stackTrace, "ALLOWED_MESSAGES",
+         ALLOWED_MESSAGES);
+    }
+
+    @Test
+    public void noHitsFromRangeQuery() throws IOException {
```
Comment (nishantnahata) [inline src/test/java/com/google/error_analyzer/backend/StackTraceTest.java:66]: {function under test}_noHitsFromRangeQuery

File: src/main/java/com/google/error_analyzer/backend/StackTrace.java [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+import org.elasticsearch.search.builder.SearchSourceBuilder;
+import org.elasticsearch.search.SearchHit;
+import org.elasticsearch.search.SearchHits;
+/**
+* Find the stack trace following an error. The algorithm fetches documents from
+* the index and see if they fit they stacktrace format.
+*/
+public class StackTrace {
+    private Integer BATCH_SIZE = 100;
+    private Integer ALLOWED_MESSAGES = 5;
```
[the remark this answers is not in the document]
Comment (Rythmag) [reply src/main/java/com/google/error_analyzer/backend/StackTrace.java:34]: Changing these variables during testing

File: src/main/java/com/google/error_analyzer/backend/StackTrace.java (hunk 1 above)
Comment (Rythmag) [reply src/main/java/com/google/error_analyzer/backend/StackTrace.java]: Updated

File: src/main/java/com/google/error_analyzer/backend/StackTrace.java [hunk 6]
```diff
@@ -0,0 +1,144 @@
...
+            if(StackTraceFormat.matchesFormat(logText)) {
+                return i;
+            }
+        }
+        logger.info("No stack found");
+        return -1;
+    }
+
+    //create list of logText from SearchHits[start, end)
+    private ImmutableList < String > extractLogTextFormHits(ImmutableList < SearchHit > hits,
```
Comment (shudhatmajain) [inline src/main/java/com/google/error_analyzer/backend/StackTrace.java:111]: extractLogText"From"Hits

File: src/main/java/com/google/error_analyzer/backend/StackTrace.java [hunk 7]
```diff
@@ -0,0 +1,144 @@
...
+        Integer startOfStack = findStartOfStack(rangeHits);
+        if (startOfStack == -1) {
+            return stackLogLinesBuilder.build();
+        } else {
+            //add message brfore stack starts to stackLogLines
+            logger.info("Adding ".concat(startOfStack.toString())
+            .concat(" messages to stack list"));
+            stackLogLinesBuilder.addAll(extractLogTextFormHits(rangeHits, 0, startOfStack));
+        }
+        ImmutableList < String > stackList = iterateHitsForFindingStack(rangeHits, startOfStack);
```
Comment (shudhatmajain) [inline src/main/java/com/google/error_analyzer/backend/StackTrace.java:57]: Can we move 57-60 inside while loop.

File: src/main/java/com/google/error_analyzer/backend/StackTrace.java [hunk 8]
```diff
@@ -0,0 +1,144 @@
...
+                return stackLogLinesBuilder.build();
+            }
+        }
+        return stackLogLinesBuilder.build();
+    }
+
+    //returns index of the first searchHit that matches the stack format
+    // -1 if start is not found
+    private Integer findStartOfStack(ImmutableList < SearchHit > hits) {
+        Integer end = Math.min(ALLOWED_MESSAGES+1, hits.size());
```
Comment (shudhatmajain) [inline src/main/java/com/google/error_analyzer/backend/StackTrace.java:98]: Should it be ALLOWED_MESSAGES or (ALLOWED_MESSAGES+1)?

File: src/main/java/com/google/error_analyzer/backend/StackTrace.java [hunk 9]
```diff
@@ -0,0 +1,144 @@
...
+import org.elasticsearch.search.builder.SearchSourceBuilder;
+import org.elasticsearch.search.SearchHit;
+import org.elasticsearch.search.SearchHits;
+/**
+* Find the stack trace following an error. The algorithm fetches documents from
+* the index and see if they fit they stacktrace format.
+*/
+public class StackTrace {
+    private Integer BATCH_SIZE = 100;
+    private Integer ALLOWED_MESSAGES = 5;
```
Comment (shudhatmajain) [inline src/main/java/com/google/error_analyzer/backend/StackTrace.java:35]: Can we have a more descriptive name?

File: src/main/java/com/google/error_analyzer/servlets/StackTraceServlet.java [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+        response.setContentType(FileConstants.APPLICATION_JSON_CONTENT_TYPE);
+        try {
+            Integer errorLineNumber = Integer.parseInt(request.getParameter(LogFields.LOG_LINE_NUMBER));
+            String indexName = request.getParameter(LogFields.FILE_NAME);
+            ImmutableList < String > stackList = stackTrace.findStack(errorLineNumber,indexName);
+            
+            if (stackList.size() == 0) {
+                String json = new Gson().toJson("No stack found");
+                response.getWriter().println(json);
+            }else{
```
Comment (shudhatmajain) [inline src/main/java/com/google/error_analyzer/servlets/StackTraceServlet.java:50]: } else {

File: src/test/java/com/google/error_analyzer/data/StackFormatTest.java [hunk 11]
```diff
@@ -0,0 +1,84 @@
...
+    @Test
+    public void startsWithAtWithSpace() {
+        String logText = " at com.stack.stacktrace.StackTraceExample.methodB";
+        boolean actual = StackTraceFormat.matchesFormat(logText);
+        boolean expected = true;
+        Assert.assertEquals(actual, expected);
+    }
+
+    @Test
+    public void AtPreceededBy20Characters() {
```
Comment (shudhatmajain) [inline src/test/java/com/google/error_analyzer/data/StackFormatTest.java:36]: Please use lowerCamelCase for function name.

File: src/main/java/com/google/error_analyzer/backend/StackTrace.java (hunk 8 above)
Comment (Rythmag) [reply src/main/java/com/google/error_analyzer/backend/StackTrace.java:98]: If we are allowing 5 messages before call stack, then at least 6 lines need to be checked

File: src/main/java/com/google/error_analyzer/backend/StackTrace.java (hunk 7 above)
Comment (Rythmag) [reply src/main/java/com/google/error_analyzer/backend/StackTrace.java:57]: Improves readability as the first case is different from the case inside while. Also it needs a different condition to check if end of stack trace was found .

File: src/main/java/com/google/error_analyzer/backend/StackTrace.java (hunk 9 above)
Comment (Rythmag) [reply src/main/java/com/google/error_analyzer/backend/StackTrace.java:35]: Updated.

File: src/main/java/com/google/error_analyzer/backend/DaoInterface.java [hunk 12]
```diff
@@ -45,4 +46,7 @@ public String storeLogLine (String filename, String jsonString, String id)
     public void bulkStoreLog(String fileName, 
         ImmutableList < Document > documentList) throws IOException;
 
+    //fetch documents from index according to searchRequest
+    public SearchHit[] getHitsFromIndex(SearchRequest searchRequest)
```
[the remark this answers is not in the document]
Comment (Rythmag) [reply src/main/java/com/google/error_analyzer/backend/DaoInterface.java]: I'll include other usages of this function while updating findAndStoreErrors in logdao

File: src/test/java/com/google/error_analyzer/data/StackFormatTest.java [hunk 13]
```diff
@@ -0,0 +1,84 @@
...
+import org.junit.Assert;
+import org.junit.runner.RunWith;
+import org.junit.runners.JUnit4;
+import org.junit.Rule;
+import org.junit.Test;
+
+public final class StackFormatTest {
+
+    @Test
+    public void startsWithAtWithSpace() {
```
Comment (nishantnahata) [inline src/test/java/com/google/error_analyzer/data/StackFormatTest.java:28]: {function under test}_

**Pull request merged**